### PR TITLE
[SPARK-15738][PYSPARK][ML] Adding Pyspark ml RFormula __str__ method similar to Scala API

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?

Adding **str** to RFormula and model that will show the set formula param and resolved formula.  This is currently present in the Scala API, found missing in PySpark during Spark 2.0 coverage review.
## How was this patch tested?

run pyspark-ml tests locally

### CODE REVIEW

File: python/pyspark/ml/feature.py [hunk 1]
```diff
@@ -2557,6 +2557,10 @@ def getFormula(self):
     def _create_model(self, java_model):
         return RFormulaModel(java_model)
 
+    def __str__(self):
+        formulaStr = self.getFormula() if self.isDefined(self.formula) else ""
```
Comment (MLnick) [inline python/pyspark/ml/feature.py]: Technically the Scala side is `Option[String]` so prints `RFormula(None) ...` when the param is not set. Not a big deal, but `... else None` here would sync it up exactly.

Comment (MLnick) [inline python/pyspark/ml/feature.py]: I think this could have a little doc test (and for the model also).

**Comment by BryanCutler** [comment]
Thanks @MLnick , I made those changes.  The string representation for the RFormulaModel is really ugly - it has lots of `WrappedArray` types.  I took all that out of the doc test, so let me know if what is there is ok.  I'm not really sure how it is even useful, maybe for debugging I guess?

**Comment by MLnick** [comment]
Yeah actually looking at it the Scala side is a bit ugly. We can make it prettier there and then just call into it from PySpark, e.g.

```
scala> val formula = new RFormula().setFormula("clicked ~ country + hour + country:hour").setFeaturesCol("features") .setLabelCol("label")
formula: org.apache.spark.ml.feature.RFormula = RFormula(clicked ~ country + hour + country:hour) (uid=rFormula_7e1422944cda)

scala> val model = formula.fit(dataset)
model: org.apache.spark.ml.feature.RFormulaModel = RFormulaModel(ResolvedRFormula(label=clicked, terms=[country,hour,{country,hour}], hasIntercept=true)) (uid=rFormula_7e1422944cda)
```
and 
```
In [8]: rf._call_java("toString")
Out[8]: u'RFormula(y ~ x + s) (uid=RFormula_495190a2bd89e76b710f)'

In [9]: model._call_java("toString")
Out[9]: u'RFormulaModel(ResolvedRFormula(label=y, terms=[x,s], hasIntercept=true)) (uid=RFormula_495190a2bd89e76b710f)'
```

This is achieved by amending `RFormula.toString` (by the way on second thought I prefer your approach of empty string instead of None):
```scala
override def toString: String = s"RFormula(${get(formula).getOrElse("")}) (uid=$uid)"
```

And overriding `ResolvedRFormula.toString`:

```scala
private[ml] case class ResolvedRFormula(
  label: String, terms: Seq[Seq[String]], hasIntercept: Boolean) {

  override def toString: String = {
    val ts = terms.map {
      case t if t.length > 1 =>
        s"${t.mkString("{", ",", "}")}"
      case t =>
        t.mkString
    }
    val termStr = ts.mkString("[", ",", "]")
    s"ResolvedRFormula(label=$label, terms=$termStr, hasIntercept=$hasIntercept)"
  }
}
```

cc @yanboliang

**Comment by yanboliang** [comment]
+1 @MLnick

**Comment by BryanCutler** [comment]
That looks pretty good to me too, thanks @MLnick! I'll put that in soon.

**Comment by BryanCutler** [comment]
@MLnick , I added the nicer string override for `ResolvedRFormula` and updated the doc tests accordingly

**Comment by MLnick** [comment]
LGTM

**Comment by BryanCutler** [comment]
Can this be merged @MLnick ?

**Comment by yanboliang** [comment]
LGTM2

**Comment by MLnick** [comment]
@yanboliang do you want to go ahead and merge this one?

**Comment by yanboliang** [comment]
I was trying to merge this, but looks like the merge script is not happy. Please feel free to merge this.

**Comment by yanboliang** [comment]
Merged to master/2.0. Thanks!

**Comment by BryanCutler** [comment]
Thanks Yanbo!

**Pull request closed without merging**